lib/partition_healing/discover_provider_healer.js

```diff
@@ -0,0 +1,169 @@
+ * 2. remove a host from the target list and try to join it;
+ * 3. merge the two membership lists when possible and gossip changes around when necessary;
+ * 4. remove the hosts that are alive according to target's membership list from the target list.
+ * 5. goto 2 until the target list is empty or the number of failures reached a configurable maximum (discoverProviderHealerMaxFailures)
+ *
+ * A full description of the algorithm is available in ringpop-common/docs.
+ *
+ * @param {Healer~healCallback} callback the callback when the heal operation is completed.
+ */
+DiscoverProviderHealer.prototype.heal = function heal(callback) {
+    var self = this;
+    self.ringpop.stat('increment', 'heal.triggered.discover_provider');
+
+    if(!self.ringpop.isReady) {
+        callback(Errors.RingpopIsNotReadyError());
+        return;
+    }
+    if (!self.ringpop.discoverProvider) {
+        var error = Errors.DiscoverProviderNotAvailableError();
+        self.logger.warn(error.message, {
+            local: self.ringpop.whoami()
+        });
+        callback(error);
+        return;
+    }
+
+    self.ringpop.discoverProvider(onHostsDiscovered);
+
+    function onHostsDiscovered(err, hosts) {
+        if (err) {
+            self.logger.warn('ringpop unable to retrieve host list from discover provider during heal', {
+                local: self.ringpop.whoami(),
+                err: err
+            });
+            return callback(err);
+        }
+
+        var potentialTargets = self._getTargets(hosts);
+        var targets = [];
+        var numberOfFailures = 0;
+
+        async.whilst(
+            function shouldContinueHealing() {
+                return potentialTargets.length > 0 && numberOfFailures < self.maxNumberOfFailures;
+            },
+            function healTargetIterator(next) {
+                var target = potentialTargets.pop();
+
+                self.attemptHeal(target, function onHealAttempt(err, reachableNodes) {
+                    if (err) {
+                        numberOfFailures++;
+                        self.logger.warn('ringpop heal attempt failed', {
+                            local: self.ringpop.whoami(),
+                            target: target,
+                            numberOfFailures: numberOfFailures,
+                            maxNumberOfFailures: self.maxNumberOfFailures,
+                            err: err
+                        });
+
+                        // continue
+                        next();
+                        return;
+                    }
+
+                    targets.push(target);
+                    // Remove all reachable nodes from the list of potential targets.
+                    potentialTargets = _.difference(potentialTargets, reachableNodes);
+                    next();
+                });
+            },
+            function healingDone(err) {
+                if (numberOfFailures >= self.maxNumberOfFailures) {
```

Comment:
There isn't a stat in ringpop-go. We might add it here but we have to add it to ringpop-go as well. @CorgiMan what do you think?